### PR TITLE
feat: storage providers settings as objects

### PR DESCRIPTION
**Pull request by tinovyatkin**
Switches Storage providers to use configuration objects rather than a single URL

### CODE REVIEW

File: extensions/storages/source/providers/Temporary.ts [hunk 1]
```diff
@@ -1,14 +1,14 @@
...
-import assert from 'node:assert'
-import { pathToFileURL } from 'node:url'
 
 import { FileSystem } from './FileSystem'
 
-export class Temporary extends FileSystem {
-  public constructor (url: URL) {
-    assert.equal(url.protocol, 'tmp:', `Invalid temporary URL: ${url.toString()}`)
+export interface TemporaryOptions {
+  prefix?: string
```
Comment (temich) [inline extensions/storages/source/providers/Temporary.ts:7]: Documentation says it should be `path`.

File: extensions/storages/source/deployment.ts [hunk 2]
```diff
@@ -16,50 +19,53 @@ export function deployment (instances: Instance[], annotation: Annotation): Depe
...
 
     for (const secret of Provider.SECRETS)
       secrets.push({
-        name: `TOA_STORAGES_${storage.toUpperCase()}_${secret.toUpperCase()}`,
+        name: `${SERIALIZATION_PREFIX}_${storage.toUpperCase()}_${secret.name.toUpperCase()}`,
         secret: {
           name: `toa-storages-${storage}`,
-          key: secret
+          key: secret.name,
+          optional: secret.optional
```
Comment (temich) [inline extensions/storages/source/deployment.ts:56]: Corresponding changes from `dev` should be merged to test this.
[This scenario](https://github.com/toa-io/toa/blob/f4faff0fb376e867fd675d192f42a70e320c7332/features/extensions/storages.feature#L78) should pass.

File: extensions/storages/source/deployment.ts [hunk 3]
```diff
@@ -16,50 +19,53 @@ export function deployment (instances: Instance[], annotation: Annotation): Depe
...
+function getSecrets (annotation: Readonly<Annotation>): Variable[] {
   const secrets: Variable[] = []
 
-  for (const [storage, ref] of Object.entries(annotation)) {
-    const url = new URL(ref)
-    const Provider = providers[url.protocol]
+  for (const [storage, props] of Object.entries(annotation)) {
+    const providerId = props.provider
+
+    validateProviderId(providerId)
```
Comment (temich) [inline extensions/storages/source/deployment.ts]: This validation has already been done at this point.

File: extensions/storages/source/deployment.ts [hunk 4]
```diff
@@ -16,50 +19,53 @@ export function deployment (instances: Instance[], annotation: Annotation): Depe
...
+
+    const Provider = providers[providerId]
 
-    if (Provider.SECRETS === undefined)
-      continue
+    if (Provider.SECRETS.length === 0) continue
```
Comment (temich) [inline extensions/storages/source/deployment.ts]: This can be removed since SECRETS is always an array.

**Review by temich — CHANGES_REQUESTED** [review]

File: extensions/storages/source/deployment.ts (hunk 3 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts]: It's a Typescript typeguard in this context.

File: extensions/storages/source/deployment.ts (hunk 4 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts]: This is the check that array is empty, but yes, can be removed as we just iterate the array later.

File: extensions/storages/source/deployment.ts (hunk 2 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts:56]: They are merged - this PR initially was on top of that changes

File: extensions/storages/source/providers/Temporary.ts (hunk 1 above)
Comment (tinovyatkin) [reply extensions/storages/source/providers/Temporary.ts:7]: Missed that, thank you! Will adjust the docs

File: extensions/storages/source/deployment.ts (hunk 2 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts:56]: Screenshot 2024-02-09 at 10 01 36

Comment (temich) [reply extensions/storages/source/deployment.ts:56]: [They](https://github.com/toa-io/toa/blob/f4faff0fb376e867fd675d192f42a70e320c7332/features/extensions/storages.feature#L106) are [not](https://github.com/toa-io/toa/blob/148a3083a29cdcfd7452cc333ace42a79322c018/features/extensions/storages.feature#L118).

File: extensions/storages/source/providers/Temporary.ts (hunk 1 above)
Comment (tinovyatkin) [reply extensions/storages/source/providers/Temporary.ts:7]: Fixed in https://github.com/toa-io/toa/pull/493/commits/8df71c999a30a660c2dca97fd31d42711656d0d5

File: extensions/storages/source/deployment.ts (hunk 3 above)
Comment (temich) [reply extensions/storages/source/deployment.ts]: Ok, I will not argue here.

But this kind of practice will not be acceptable on hot code paths.

We cannot run actual code in production that only exists to satisfy meta-code requirements on developers' machines.

File: extensions/storages/source/deployment.ts (hunk 4 above)
Comment (temich) [reply extensions/storages/source/deployment.ts]: ~Yes, and this check that an array is empty can be removed.~

File: extensions/storages/source/deployment.ts (hunk 2 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts:56]: Screenshot 2024-02-09 at 10 12 46

Rebased. Test still passing.

Comment (temich) [reply extensions/storages/source/deployment.ts:56]: I looked at the wrong scenario, sorry.

File: extensions/storages/source/deployment.ts (hunk 4 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts]: Removed in https://github.com/toa-io/toa/pull/493/commits/6cc7026fb95dd765addabc2d3d30712d00455fe8

File: extensions/storages/source/deployment.ts (hunk 3 above)
Comment (tinovyatkin) [reply extensions/storages/source/deployment.ts]: Fixed in https://github.com/toa-io/toa/pull/493/commits/6cc7026fb95dd765addabc2d3d30712d00455fe8

**Review by temich — APPROVED** [review]

**Pull request merged**